unimod_mapper/unimod_mapper.py

```diff
@@ -688,6 +719,165 @@ def _reparseXML(self, xml_file_list=[]):
+        the mapping. Moreover, it can only have the minimal amound of items, or e.g.
+        engine-specific ones. At the end the mapped vallues will be updated to the
+        original dict.
+
+        Args:
+            mod_list (list): list of mod_dicts containing all relevant info about a
+                             given modification
+
+        Returns:
+            rdict (dict): dict with mod types as keys and corresponding lists of
+                             mod dicts mapped to unimod
+
+        Examples:
+
+            mod_list = [
+                {
+                    "aa": "M",              # specify the modified amino acid as a single letter, use '*' if the amino acid is variable
+                    "type": "opt",          # specify if it is a fixed (fix) or potential (opt) modification
+                    "position": "any",      # specify the position within the protein/peptide (Prot-N-term, Prot-C-term), use 'any' if the positon is variable
+                    "name": "Oxidation",    # specify the unimod PSI-MS Name (alternative to id)
+                    "id": None,             # specify the unimod Accession (alternative to name)
+                    "composition": None,    # For user-defined mods composition needs to be given as a Hill notation
+                }
+            ]
+
+        """
+
+        rdict = {"fix": [], "opt": []}
+        for index, mod in enumerate(mod_list):
+
+            # Check if any key from the mod dict is not valid!
+            allowed_keys = ["aa", "type", "position", "name", "id", "composition",
+                            "max_num_per_peptide", "intern_dist", "required", "neutral_loss"]
+            wrong_keys = []
+            for key in mod.keys():
+                if key not in allowed_keys:
+                    wrong_keys.append(key)
+            if wrong_keys != []:
+                logger.warning(f"One or more used keys are not supported. Please "
+                             f"confirm that you are using keys from {allowed_keys} "
+                             f"within your mod_dict. Continue without {mod}!")
+                # print(mod)
+                continue
+
+            unimod = False
+            unimod_id = None
+            type = mod["type"]
+
+            if mod.get("composition", None) is None:
+                try:
+                    unimod_id = int(mod["id"])
+                    unimod_name = self.id2name_list(unimod_id)
+                    mass = self.id2mass_list(unimod_id)
+                    composition = self.id2composition_list(unimod_id)
+                    if unimod_name == []:
+                        logger.warning(
+                            "'{1}' is not a Unimod modification please change it to a valid Unimod Accession # or PSI-MS Unimod Name or add the chemical composition hill notation (including 1) e.g.: H-1N1O2 ursgal_style: 'amino_acid,opt/fix,position,name,chemical_composition'. Continue without modification {0} ".format(
```

Comment:
"or add the chemical composition hill notation (including 1) e.g.: H-1N1O2 ursgal_style: 'amino_acid,opt/fix,position,name,chemical_composition'"
--> or add the chemical composition as hill notation to the mod_dict, e.g:
{"composition": "H-1N1O2"}

Reply:
not sure if we need to support this ...

Reply:
I changed the phrasing, but happy to remove it completely if it is not longer needed?

Reply:
@fu How else would you add userdefined mods if we don't support it here?